esmvaltool/diag_scripts/weighting/climwip.py

```diff
@@ -371,12 +372,71 @@ def compute_overall_mean(dataset: 'xr.Dataset',
     return overall_mean
 
 
+def combine_dimension(dataset: 'xr.DataArray', dimn: str) -> 'xr.DataArray':
+    """Combine ensemble members of the same model along the given dimension."""
+    model_ensemble = dataset[dimn].values
+    models = [me.split('_')[0] for me in model_ensemble]
+    dataset['model'] = xr.DataArray(models, dims=dimn)
+    dataset = dataset.groupby('model').mean(keep_attrs=True).rename(
+        {'model': dimn})
+
+    # new model names
+    groups = {
+        k: [*v]
+        for k, v in itertools.groupby(model_ensemble,
+                                      lambda x: x.split('_')[0])
+    }
+    for key, group in groups.items():
+        model, forcing = group[0].split('_')[::2]
+        members = '-'.join([me.split('_')[1] for me in group])
+        groups[key] = '_'.join([model, members, forcing])
+
+    # ensure that they are sorted correctly
+    return dataset.assign_coords(
+        {dimn: [groups[key] for key in dataset[dimn].values]})
+
+
+def combine_ensemble_members(
+        dataset: Union['xr.DataArray', None]) -> Union['xr.DataArray', None]:
+    """Combine ensemble members of the same model along all dimensions."""
+    if dataset is None:
+        return None
+    dataset = combine_dimension(dataset, 'model_ensemble')
+    if 'perfect_model_ensemble' in dataset.dims:
+        dataset = combine_dimension(dataset, 'perfect_model_ensemble')
+        # need to set the diagonal elements back to zero after averaging
+        dataset.values[np.diag_indices(dataset['model_ensemble'].size)] = 0
```

Comment:
This seems a weird artefact due to the fact that you're averaging over 2 dimensions after one another, where the length of the dimensions changes in between. Doesn't this affect the other values as well?

I wonder if it is possible and perhaps neater to groupby a multi-dimensional array in this case.

Reply:
See my comment above on the models DataArray.

Reply:
Then you also don't need to split the function in 2.

Reply:
Yes I always found this not so nice as well but I'm not sure if it can be resolved any other way. A model has 0 distance to itself but several ensemble members do not have 0 distance to themselfs in the mean. This is they way it is with internal variability present. The diagonal elements are basically a measure for internal variability that we disregard here because we do not need it. 

It does also affect the other elements but this is a design choice. The other option would be to just use one model per member in the perfect model dimension but I think this would make it even less intuitive and less robust.

Reply:
Ah yes, okay. I now get why the diagonal is not zero. But still I'm a bit concerned about the other values, especially seeing that the result is not symmetric anymore, and shouldn't it be? So I've been playing around with a simple example to see if I can create a multidimensional grouping. It can still be polished, but at this point I just wonder what do you think?

```python
ds = xr.DataArray([[0, 2, 3],[1, 0, 3],[1, 2, 0]], dims=['x', 'y'], coords={'x': ['a', 'b', 'b'], 'y': ['a', 'b', 'b']})
print(ds)

> <xarray.DataArray (x: 3, y: 3)>
> array([[0, 2, 3],
>        [1, 0, 3],
>        [1, 2, 0]])
> Coordinates:
>   * x        (x) <U1 'a' 'b' 'b'
>   * y        (y) <U1 'a' 'b' 'b'

# Groupby 1 dimension at a time
modelx = xr.DataArray(['a','b','b'], dims=['x'], coords={'x': ['a', 'b', 'b']})
modely = xr.DataArray(['a','b','b'], dims=['y'], coords={'y': ['a', 'b', 'b']})
method1 = ds.groupby(modelx).mean().rename({'group': 'x'}).groupby(modely).mean().rename({'group': 'y'})
print(method1)  # Notice that the result is not symmetric

> <xarray.DataArray (x: 2, y: 2)>
> array([[0.  , 2.5 ],
>        [1.  , 1.25]])
> Coordinates:
>   * x        (x) object 'a' 'b'
>   * y        (y) object 'a' 'b'

# Now make multidimensional groups
groups = np.char.add(*np.meshgrid(modelx.values, modely.values))
resolve = lambda x: ''.join(sorted(list(set(x))))  # 'ba' --> 'ab'
groups = np.vectorize(resolve)(groups)
modelxy = xr.DataArray(groups, dims=['x', 'y'], coords={'x': ['a', 'b', 'b'], 'y': ['a', 'b', 'b']})
print(modelxy)

> <xarray.DataArray (x: 3, y: 3)>
> array([['a', 'ab', 'ab'],
>        ['ab', 'b', 'b'],
>        ['ab', 'b', 'b']], dtype='<U2')
> Coordinates:
>   * x        (x) <U1 'a' 'b' 'b'
>   * y        (y) <U1 'a' 'b' 'b'

method2 = ds.groupby(modelxy).mean()
print(method2)  # note that 1.75 is the mean of 1 and 2.5, is that a coincidence?

> <xarray.DataArray (group: 3)>
> array([0.  , 1.75, 1.25])
> Coordinates:
>   * group    (group) object 'a' 'ab' 'b'
```

Reply:
Hi Peter!
I haven't looked at this in detail but the input matrix does not look symmetric to me?

> ``ds = xr.DataArray([[0, 2, 3],[1, 0, 3],[1, 2, 0]], dims=['x', 'y'], coords={'x': ['a', 'b', 'b'], 'y': ['a', 'b', 'b']})

If I do this, the output is the same for both your cases
``ds = xr.DataArray([[0, 2, 3],[2, 0, 1],[3,1, 0]], dims=['x', 'y'], coords={'x': ['a', 'b', 'b'], 'y': ['a', 'b', 'b']})

Reply:
Ahh, you're right. So we get away with it after all. I remember a colleague of mine had bad experience with a similar code once, so that's why I was so cautious. But I think it's okay then.